### PR TITLE
AA8K-17 

### PR DESCRIPTION
**Pull request by Bartek-Figat**
Write in typescript route /auth-with-google using passport google strategy to authenticate with google oauth.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -14,14 +14,19 @@
...
   },
   "devDependencies": {
     "@commitlint/cli": "^9.1.2",
     "@types/cors": "^2.8.7",
     "@types/jest": "^26.0.13",
     "@types/mongoose": "^5.7.36",
     "@types/node": "^14.6.4",
+    "@types/passport": "^1.0.4",
+    "@types/passport-google-oauth": "^1.0.41",
+    "@types/passport-google-oauth2": "^0.1.3",
```
Comment (hayuna) [inline package.json]: What is the difference and do we need both?

File: src/models/User.d.ts [hunk 2]
```diff
@@ -1,22 +1,25 @@
...
 
-interface IUserDoc extends Document, IUser {
-}
+interface IUserDoc extends Document, IUser {}
 
 interface IUserModel extends Model<IUserDoc> {
-  findOneOrCreate(accessToken: string, profile: IProfile): Promise<IUserDoc>
+  findOrCreate(
+    arg0: { googleId: string },
+    arg1: (err: any, user: any) => any
```
Comment (hayuna) [inline src/models/User.d.ts]: `any` type is not acceptable.
`arg0`, `arg1` are you sure this is the best variable name?

File: src/routes/auth.ts [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+});
+
+passport.deserializeUser((user, cb) => {
+  cb(null, user);
+});
+
+passport.use(
+  new OAuthStrategy(
+    {
+      consumerKey: "GOOGLE_CONSUMER_KEY",
```
Comment (hayuna) [inline src/routes/auth.ts]: Please move it to `.env`

File: src/routes/auth.ts [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+);
+
+router.get(
+  "/auth/google",
+  passport.authenticate("google", { scope: ["profile", "email"] })
+);
+
+router.get(
+  "/auth/google/callback",
+  passport.authenticate("google", { failureRedirect: "/login" }),
```
Comment (hayuna) [inline src/routes/auth.ts]: Hmm.. Can we use `failureRedirect` and `successRedirect` here?

**Review by hayuna — CHANGES_REQUESTED** [review]
Comments above

File: src/models/User.d.ts (hunk 2 above)
Comment (Bartek-Figat) [reply src/models/User.d.ts]: Do you have any particular sugestion for naming?

File: src/routes/auth.ts (hunk 3 above)
Comment (Bartek-Figat) [reply src/routes/auth.ts]: Sure.

File: src/routes/auth.ts (hunk 4 above)
Comment (Bartek-Figat) [reply src/routes/auth.ts]: Actually, we don't need this.

File: package.json (hunk 1 above)
Comment (Bartek-Figat) [reply package.json]: the diffrerence berween both libraries is   
oAuth 1.0 accept  consumer key, consumer secret, and callback URL
OAuth 2.0 accept  client ID, client secret, and callback URL, i will delete @types/passport-google-oauth2.

File: src/routes/auth.ts [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+});
+
+passport.deserializeUser((user, cb) => {
+  cb(null, user);
+});
+
+passport.use(
+  new OAuthStrategy(
+    {
+      consumerKey: consumerKey,
```
Comment (hayuna) [inline src/routes/auth.ts:27]: If we have the same name for key and value in object, we can shorthand it

```suggestion
      consumerKey,
```

File: src/routes/auth.ts [hunk 6]
```diff
@@ -43,7 +51,7 @@ router.get(
 
 router.get(
   "/auth/google/callback",
-  passport.authenticate("google", { failureRedirect: "/login" }),
+  passport.authenticate("google"),
```
Comment (hayuna) [inline src/routes/auth.ts:54]: You missed me. I mean that we can use two cases:
- failureRedirect
- successRedirect

File: src/routes/auth.ts (hunk 4 above)
Comment (hayuna) [reply src/routes/auth.ts]: Could you argument it?

**Pull request closed without merging**